activerecord/lib/active_record/attribute_methods/write.rb

```diff
@@ -36,8 +36,7 @@ def write_attribute(attr_name, value)
         end
 
         name = self.class.primary_key if name == "id".freeze && self.class.primary_key
-        @attributes.write_from_user(name, value)
-        value
+        _write_attribute(name, value)
       end
 
       def raw_write_attribute(attr_name, value) # :nodoc:
```

Comment:
Before https://github.com/rails/rails/pull/29464 the two methods were [branches inside `write_attribute_with_type_cast`](https://github.com/rails/rails/blob/v5.1.1/activerecord/lib/active_record/attribute_methods/write.rb#L55-L59), which I liked even less, but I agree it's not great.

Maybe instead of having a separate method, we could decorate the value with a "don't typecast me" wrapper, and push the conditional into the AttributeSet class?